ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/MergeCompactor.java

```diff
@@ -0,0 +1,210 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+
+final class MergeCompactor extends QueryCompactor {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MergeCompactor.class.getName());
+
+  @Override
+  public boolean run(HiveConf hiveConf, Table table, Partition partition, StorageDescriptor storageDescriptor,
+                  ValidWriteIdList writeIds, CompactionInfo compactionInfo, AcidDirectory dir) throws IOException, HiveException, InterruptedException {
+    if (isMergeCompaction(hiveConf, dir, writeIds, storageDescriptor)) {
+      // Only inserts happened, it is much more performant to merge the files than running a query
+      Path outputDirPath = getCompactionOutputDirPath(hiveConf, writeIds,
+              compactionInfo.isMajorCompaction(), storageDescriptor);
+      try {
+        return mergeOrcFiles(hiveConf, compactionInfo.isMajorCompaction(),
+                dir, outputDirPath, AcidUtils.isInsertOnlyTable(table.getParameters()));
+      } catch (Throwable t) {
+        // Error handling, just delete the output directory,
+        // and fall back to query based compaction.
+        FileSystem fs = outputDirPath.getFileSystem(hiveConf);
+        if (fs.exists(outputDirPath)) {
+          fs.delete(outputDirPath, true);
+        }
+        return false;
+      }
+    } else {
+      return false;
+    }
+  }
+
+  /**
+   * Returns whether merge compaction must be enabled or not.
+   * @param conf Hive configuration
+   * @param directory the directory to be scanned
+   * @param validWriteIdList list of valid write IDs
+   * @param storageDescriptor storage descriptor of the underlying table
+   * @return true, if merge compaction must be enabled
+   */
+  private boolean isMergeCompaction(HiveConf conf, AcidDirectory directory,
+                                   ValidWriteIdList validWriteIdList,
+                                   StorageDescriptor storageDescriptor) {
+    return conf.getBoolVar(HiveConf.ConfVars.HIVE_MERGE_COMPACTION_ENABLED)
+            && storageDescriptor.getOutputFormat().equalsIgnoreCase("org.apache.hadoop.hive.ql.io.orc.OrcOutputFormat")
+            && !hasDeleteOrAbortedDirectories(directory, validWriteIdList);
+  }
+
+  /**
+   * Scan a directory for delete deltas or aborted directories.
+   * @param directory the directory to be scanned
+   * @param validWriteIdList list of valid write IDs
+   * @return true, if delete or aborted directory found
+   */
+  private boolean hasDeleteOrAbortedDirectories(AcidDirectory directory, ValidWriteIdList validWriteIdList) {
+    if (!directory.getCurrentDirectories().isEmpty()) {
+      final long minWriteId = validWriteIdList.getMinOpenWriteId() == null ? 1 : validWriteIdList.getMinOpenWriteId();
+      final long maxWriteId = validWriteIdList.getHighWatermark();
+      return directory.getCurrentDirectories().stream()
+              .filter(AcidUtils.ParsedDeltaLight::isDeleteDelta)
```

Comment:
should we only include aborted deletes? I don't quite understand the check `!directory.getAbortedDirectories().isEmpty()` could you please elaborate here